JSON cmdline interface

peterschrammel opened the pull request:
Allows reading the commandline in JSON format from stdin

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

peterschrammel committed:
Add JSON cmdline interface

allows reading the command line in JSON format
from stdin.

peterschrammel committed:
Add jsont::is_boolean

return true if a property is either true or false.

peterschrammel committed:
Add json-interface to CBMC

Allows reading the commandline in JSON format
from stdin.

tautschnig commented on src/json/json_interface.cpp:
```diff
@@ -0,0 +1,101 @@
+
+#include "json_interface.h"
+
+#include <iostream>
```
Nit pick: I'd prefer for standard-library includes to come last, see also some recent exchange that @kroening and I had; the idea is that any included header file should parse even without additional includes being present. So do the most minimal ones (the ones from `util`) first, and then by increasing complexity.

tautschnig commented on src/json/json_interface.cpp:
```diff
@@ -0,0 +1,101 @@
+    throw invalid_command_line_argument_exceptiont(
+      "array expected", "'options'");
+  }
+  for(const auto &option_pair : to_json_object(options))
```
Nit pick: add a blank line before this one to be consistent with the code a few lines above.

tautschnig commented on src/json/json_interface.cpp:
```diff
@@ -0,0 +1,101 @@
+  {
+    if(argument.is_string())
+      cmdline.args.push_back(argument.value);
+    else
```
I believe you could increase readability by doing
```
if(!argument.is_string())
{
  throw ...
}

cmdline.args.push_back(argument.value);
```
because it reduces the number of closing braces that immediately follow each other. This applies even more so down below, where you currently have three closing braces.

tautschnig commented on src/json/json_interface.cpp:
```diff
@@ -0,0 +1,101 @@
+  }
+}
+
+/// Parse JSON-formatted commandline options from stdin
```
I think this should just be removed - or at least I have no idea what Doxygen makes of documentation at both declaration and definition when they aren't the same.

tautschnig commented on src/json/json_interface.h:
```diff
@@ -0,0 +1,20 @@
+
+class cmdlinet;
+
+/// Parses the JSON-formatted command line from stdin
```
I think it would be rather important to document the expected JSON syntax! Otherwise you're bound to trial and error...

tautschnig commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -1030,6 +1033,7 @@ void cbmc_parse_optionst::help()
     " --xml-ui                     use XML-formatted output\n"
     " --xml-interface              bi-directional XML interface\n"
     " --json-ui                    use JSON-formatted output\n"
+    " --json-interface             bi-directional JSON interface\n"
```
Is this actually bi-directional?! Doesn't it only become "bi"directional when using both `--json-interface` and `--json-ui`? (The same is of course true for XML.)

peterschrammel replied:
I was wondering about that and have now more investigated in the XML case (see comments there). The original intention was that --xml-interface subsumes --xml-ui. 
Actually, the current behaviour was broken because the UI was set before the option could have been parsed from stdin, which seems difficult to fix without splitting the `parse_optionst` constructor into two phases and adding a `set_ui` to `ui_message_handlert` (rather no).
Also, the question is what happens with errors during parsing the options - they should already be output in a proper format. Thus, I'd rather stick with the original intention of 'bi-directional',

tautschnig replied:
I'm not sure I fully understand - I agree that `--{json,xml}-interface` should subsume `--{json,xml}-ui` - but I don't think that was actually work as such? If things can be made to work as such then we'd have the desired bi-directional interface.

peterschrammel replied:
`--json-interface` now implies `--json-ui`, and `--xml-interface` now implies `--xml-ui`.

>but I don't think that was actually work as such?

Not sure, I understand...

tautschnig replied:
Let me retry, my apologies for completely broken English:

1) A "bi"-directional interface is about input _and_ output.
2) `--{json,xml}-ui` set up the _output_ side of an interface.
3) The comment claims that `--{json,xml}-interface` configure a bi-directional interface.
4) The actual implementation, however, only reads _input_ when using `--{json,xml}-interface`.
5) One possible fix would be that upon seeing `--{json,xml}-interface` we also set the option for `--{json,xml}-ui`.
6) This might not actually work as it could be too late, in which case a more extensive change would be required. If this is the case, a simple fix would be to change the comment such as not to promise a bi-directional interface. Users wanting to use a bi-directional interface would then need to provide both `--{json,xml}-interface` _and_ `--{json,xml}-ui`.

peterschrammel replied:
1. yes
2. yes
3. yes
4. yes
5. is being fixed here (was broken)
6. not required because 5 works.

peterschrammel committed:
Use JSON_UI with json-interface

json-interface subsumes json-ui.
When json-interface is enabled then other
commandline options haven't been parsed yet
when the UI is set.
